pkg/sql/internal.go

```diff
@@ -1158,3 +1244,45 @@ func (ief *InternalExecutorFactory) NewInternalExecutor(
 	ie.SetSessionData(sd)
 	return &ie
 }
+
+// NewInternalExecutorWithTxn creates an internal executor with txn-related info,
+// such as descriptor collection and schema change job records, etc. It should
+// be called only after InternalExecutorFactory.NewInternalExecutor is already
+// called to construct the InternalExecutorFactory with required server info.
+// This function should only be used under CollectionFactory.TxnWithExecutor().
+func (ief *InternalExecutorFactory) NewInternalExecutorWithTxn(
+	sd *sessiondata.SessionData, sv *settings.Values, txn *kv.Txn, descCol *descs.Collection,
+) (sqlutil.InternalExecutor, sqlutil.InternalExecutorCommitTxnFunc) {
+	schemaChangeJobRecords := make(map[descpb.ID]*jobs.Record)
+	// By default, if not given session data, we initialize a sessionData that
+	// would be the same as what would be created if root logged in.
+	// The sessionData's user can be override when calling the query
+	// functions of internal executor.
+	// TODO(janexing): since we can be running queries with a higher privilege
+	// than the actual user, a security boundary should be added to the error
+	// handling of internal executor.
+	if sd == nil {
+		sd = NewFakeSessionData(sv)
+		sd.UserProto = username.RootUserName().EncodeProto()
+	}
+	ie, commitTxnFunc := newInternalExecutorWithTxn(
+		ief.server,
+		sd,
+		txn,
+		ief.memMetrics,
+		ief.monitor,
+		descCol,
+		schemaChangeJobRecords,
+	)
+
+	return ie, commitTxnFunc
+}
+
+// RunWithoutTxn is to create an internal executor without binding to a txn,
+// and run the passed function with this internal executor.
+func (ief *InternalExecutorFactory) RunWithoutTxn(
```

Comment:
Why was this method necessary / how is it different from `NewInternalExecutor()` ? Can't the caller run exactly the two lines inside this function?
The comment talks about "binding to a txn" without other explanations. I don't think either fewer or more words would be needed since most callers don't care about that / don't know what "binding to a txn" means.

Reply:
I agree that this function is far from ideal... We made this function *with the hope* of avoiding use cases where an internal executor is created without binding to any txn-related metadata, but is used to run queries with a not nil-txn. In other words, we wanted to make the usages "with" and "without" an outer txn more distinct from each other, and let callers think twice about which one they should use.
I think we can add a comment saying that it's disallowed to use this function to run DDLs or multiple statement in a transactional manner.

Reply:
You can have a `NewInternalExecutorWithoutTxn` if you insist on the importance of having the "without part" in your face, but I don't see why the caller needs to structure it's logic into a closure if they're not getting anything in return.

Reply:
I think if we only have `NewInternalExecutorWithoutTxn`, it can still happen to user to use it with a not-nil txn, which is wrong. To wrap it in this function is to make it more explicit that you shouldn't do this (though it's true that we can't truly disallow it here)

I think the ideal case is to remove the `txn` field in internal executor's query functions (e.g. `ie.QueryRowEx()`). The txn should be bound to the internal executor, rather than each statement execution. With that, I think it's totally fine for us to remove this function and just do `ie := NewInternalExecutorWithoutTxn ()`